index.js

```diff
@@ -137,11 +138,58 @@ class Installer {
           this.pkgCount++
           return this
         })
+        .catch(e => {
+          if (child.optional) {
+            this.pkgCount++
+            child.failed = true
+            return rimraf(childPath)
+          } else {
+            throw e
+          }
+        })
       })
       return child.pending
     }, { concurrency: 50 })
   }
 
+  // A cute little mark-and-sweep collector!
+  garbageCollect (tree) {
+    const liveDeps = new Set()
+    const installer = this
+    mark(tree)
+    return sweep(tree)
+
+    function mark (tree) {
+      for (let dep of tree.dependencies.values()) {
+        if (dep.seenByMark) { continue }
+        dep.seenByMark = true
+        if (!dep.optional || !dep.failed) {
+          liveDeps.add(dep)
+          mark(dep)
+        }
+      }
+    }
+
+    function sweep (tree) {
+      return BB.map(tree.dependencies.values(), dep => {
+        if (dep.seenBySweep) { return }
+        dep.seenBySweep = true
+        return sweep(dep).then(() => {
+          if (!liveDeps.has(dep) && !dep.purged) {
```

Comment:
If there was a failed Set as suggested above, you could invert this conditional as `if (failedDeps.has(dep))`

Reply:
how would you mark transitive deps that didn't themselves fail?